src/cmd/tools/verify_data_files/main/main.go

```diff
@@ -0,0 +1,473 @@
+}
+
+func fixFileSet(
+	opts verifyFileSetOptions,
+	log *zap.Logger,
+) error {
+	fsOpts := fs.NewOptions().SetFilePathPrefix(opts.filePathPrefix)
+	reader, err := fs.NewReader(opts.bytesPool, fsOpts)
+	if err != nil {
+		return err
+	}
+
+	fileSet := opts.fileSet
+
+	openOpts := fs.DataReaderOpenOptions{
+		Identifier:  fileSet.ID,
+		FileSetType: persist.FileSetFlushType,
+	}
+
+	err = reader.Open(openOpts)
+	if err != nil {
+		return err
+	}
+
+	defer reader.Close()
+
+	// NOTE: we output to a new directory so that we don't clobber files.
+	writeFsOpts := fsOpts.SetFilePathPrefix(opts.fixDir)
+	writer, err := fs.NewWriter(writeFsOpts)
+	if err != nil {
+		return err
+	}
+
+	err = writer.Open(fs.DataWriterOpenOptions{
+		FileSetType:        persist.FileSetFlushType,
+		FileSetContentType: fileSet.ID.FileSetContentType,
+		Identifier:         fileSet.ID,
+		BlockSize:          reader.Status().BlockSize,
+	})
+	if err != nil {
+		return err
+	}
+
+	var (
+		currTags     []ident.Tag
+		currTagsIter = ident.NewTagsIterator(ident.Tags{})
+		tagsBuffer   []ident.Tag
+		removedIDs   int
+		removedTags  int
+		copies       []checked.Bytes
+	)
+	for {
+		id, tags, data, checksum, err := reader.Read()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return err
+		}
+
+		// Need to save tags in case we need to write them out again
+		// (iterating them in read entry means we can't reiterate them
+		// without copying/duplicating).
+		currTags = currTags[:0]
+		for tags.Next() {
+			tag := tags.Current()
+			name := tag.Name.Bytes()
+			value := tag.Value.Bytes()
+
+			// Need to take copy as only valid during iteration.
+			nameCopy := opts.bytesPool.Get(len(name))
+			nameCopy.IncRef()
+			nameCopy.AppendAll(name)
+			valueCopy := opts.bytesPool.Get(len(value))
+			valueCopy.IncRef()
+			valueCopy.AppendAll(value)
+			copies = append(copies, nameCopy)
+			copies = append(copies, valueCopy)
+
+			currTags = append(currTags, ident.Tag{
+				Name:  ident.BytesID(nameCopy.Bytes()),
+				Value: ident.BytesID(valueCopy.Bytes()),
+			})
+		}
+
+		// Choose to write out the current tags if do not need modifying.
+		writeTags := currTags[:]
+
+		var currIdentTags ident.Tags
+		currIdentTags.Reset(currTags)
+		currTagsIter.Reset(currIdentTags)
+
+		check, err := readEntry(id, currTagsIter, data, checksum)
+		if err != nil {
+			shouldFixInvalidID := check.invalidID && opts.fixInvalidIDs
+			if shouldFixInvalidID {
+				// Skip this entry being written to the target volume.
+				removedIDs++
+				continue
+			}
+
+			shouldFixInvalidTags := check.invalidTags && opts.fixInvalidTags
+			if shouldFixInvalidTags {
+				// Need to remove invalid tags.
+				tagsBuffer = tagsBuffer[:0]
+				for _, tag := range currTags {
+					if err := convert.ValidateSeriesTag(tag); err != nil {
+						removedTags++
+						continue
+					}
+					tagsBuffer = append(tagsBuffer, tag)
+				}
+
+				// Make sure we write out the modified tags.
+				writeTags = tagsBuffer
```

Comment:
Doesn't this risk 2 series with the same tagset? Should probably skip the entire series instead IMO

Reply:
Sure thing, can remove the series.